core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -885,7 +921,7 @@ private Builder(TableMetadata base) {
       this.sortOrders = Lists.newArrayList(base.sortOrders);
       this.properties = Maps.newHashMap(base.properties);
       this.currentSnapshotId = base.currentSnapshotId;
-      this.snapshots = Lists.newArrayList(base.snapshots);
+      this.snapshots = Lists.newArrayList(base.snapshots());
```

Comment:
Is this needed?

It will always load snapshots when building a new table metadata and that doesn't seem right. I think it would be correct to do it that way, but you don't really need to load them as long as the `snapshotsSupplier` is passed here as well. If all of the snapshots are needed to write the new metadata to a file, the metadata parser will call `snapshots()` at which point they will be loaded.

The only problem I can think of is that we're replacing the snapshot list completely in `ensureSnapshotsLoaded` and we would need to update that to basically union the initially loaded snapshots with the new ones that are before the sequence number. That's a bit more complex and we could do that later if we wanted to.

I think the argument for lazily loading snapshots is that this is in the critical path during a commit, so hitting a REST endpoint is going to increase the chances of conflicts. On the other hand, this is a simpler way to make snapshot loading lazy for all read cases.

What do you think?

Reply:
For now, it's safest to eagerly load snapshots when building new metadata.  The issue with carrying over the snapshot supplier is that we don't perform the proper validation and pruning unless we extract the logic that is done in the base metadata.

I think that would be a good follow up to see if it's safe, but I don't think we need to do that immediately.